src/pluggy/_manager.py

```diff
@@ -154,46 +155,80 @@ def register(self, plugin: _Plugin, name: str | None = None) -> str | None:
 
         # register matching hook implementations of the plugin
         for name in dir(plugin):
-            hookimpl_opts = self.parse_hookimpl_opts(plugin, name)
-            if hookimpl_opts is not None:
-                normalize_hookimpl_opts(hookimpl_opts)
+            hookimpl_config = self._parse_hookimpl(plugin, name)
+            if hookimpl_config is not None:
                 method: _HookImplFunction[object] = getattr(plugin, name)
-                hookimpl = HookImpl(plugin, plugin_name, method, hookimpl_opts)
-                name = hookimpl_opts.get("specname") or name
-                hook: HookCaller | None = getattr(self.hook, name, None)
+                hookimpl = HookImpl(plugin, plugin_name, method, hookimpl_config)
+                hook_name = hookimpl_config.specname or name
+                hook: HookCaller | None = getattr(self.hook, hook_name, None)
                 if hook is None:
-                    hook = HookCaller(name, self._hookexec)
-                    setattr(self.hook, name, hook)
+                    hook = HookCaller(hook_name, self._hookexec)
+                    setattr(self.hook, hook_name, hook)
                 elif hook.has_spec():
                     self._verify_hook(hook, hookimpl)
                     hook._maybe_apply_history(hookimpl)
                 hook._add_hookimpl(hookimpl)
         return plugin_name
 
+    def _parse_hookimpl(
+        self, plugin: _Plugin, name: str
+    ) -> HookimplConfiguration | None:
+        """Internal method to parse hook implementation configuration.
+
+        This method uses the new HookimplConfiguration type internally.
+        Falls back to the legacy parse_hookimpl_opts method for compatibility.
+
+        :param plugin: The plugin object to inspect
+        :param name: The attribute name to check for hook implementation
+        :returns: HookimplConfiguration if found, None otherwise
+        """
+        try:
+            method: object = getattr(plugin, name)
+        except Exception:  # pragma: no cover
+            return None
+
+        if not inspect.isroutine(method):
+            return None
+
+        try:
+            # Try to get hook implementation configuration directly
+            impl_attr = getattr(method, self.project_name + "_impl", None)
+        except Exception:  # pragma: no cover
+            impl_attr = None
+
+        if impl_attr is not None:
```

Comment:
this one is a artifact i still need to resolve, but the followup on those is much lighter than the heavy lifting before